lib/Zonemaster/Backend/Validator.pm

```diff
@@ -7,36 +7,83 @@ use warnings;
 use 5.14.2;
 
 use Exporter qw( import );
+use File::Spec::Functions qw( file_name_is_absolute );
 use JSON::Validator::Joi;
 use Readonly;
 
 our @EXPORT_OK = qw(
+  untaint_abs_path
+  untaint_ldh_domain
   untaint_engine_type
+  untaint_mariadb_database
+  untaint_mariadb_user
+  untaint_password
+  untaint_positive_int
+  untaint_positive_millis
+  untaint_postgresql_ident
+  untaint_unsigned_int
 );
 
 our %EXPORT_TAGS = (
     untaint => [
         qw(
+          untaint_abs_path
+          untaint_ldh_domain
           untaint_engine_type
+          untaint_mariadb_database
+          untaint_mariadb_user
+          untaint_password
+          untaint_positive_int
+          untaint_positive_millis
+          untaint_postgresql_ident
+          untaint_unsigned_int
           )
     ],
 );
 
+# Does not check value ranges within the groups
 Readonly my $IPV4_RE => qr/^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}$/;
+
+# Does not check the length and number of the hex groups, nor the value ranges in the IPv4 groups
 Readonly my $IPV6_RE => qr/^([0-9a-f]{1,4}:[0-9a-f:]{1,}(:[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})?)$|([0-9a-f]{1,4}::[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})$/i;
 
-Readonly my $API_KEY_RE             => qr/^[a-z0-9-_]{1,512}$/i;
-Readonly my $CLIENT_ID_RE           => qr/^[a-z0-9-+~_.: ]{1,50}$/i;
-Readonly my $CLIENT_VERSION_RE      => qr/^[a-z0-9-+~_.: ]{1,50}$/i;
-Readonly my $DIGEST_RE              => qr/^[a-f0-9]{40}$|^[a-f0-9]{64}$/i;
-Readonly my $ENGINE_TYPE_RE         => qr/^(?:mysql|postgresql|sqlite)$/i;
-Readonly my $IPADDR_RE              => qr/^$|$IPV4_RE|$IPV6_RE/;
-Readonly my $JSONRPC_METHOD_RE      => qr/^[a-z0-9_-]*$/i;
-Readonly my $LANGUAGE_RE            => qr/^[a-z]{2}(_[A-Z]{2})?$/;
+Readonly my $API_KEY_RE                 => qr/^[a-z0-9-_]{1,512}$/i;
+Readonly my $CLIENT_ID_RE               => qr/^[a-z0-9-+~_.: ]{1,50}$/i;
+Readonly my $CLIENT_VERSION_RE          => qr/^[a-z0-9-+~_.: ]{1,50}$/i;
+Readonly my $DIGEST_RE                  => qr/^[a-f0-9]{40}$|^[a-f0-9]{64}$/i;
```

Comment:
Thanks for looking it up! I'll add the 96 digit case as well.

Reply:
In one way it is good that the digest is validated, but the problem is that the error message from RPCAPI is not shown to the user in the GUI, and even if it did it would be hard to understand it for most users. I removed the last character in a valid hash and sent it to zmb:

```
"error":{
  "message":"Invalid method parameter(s).",
  "code":"-32602",
  "data":"/ds_info/0/digest: String does not match (?^ui:^[a-f0-9]{40}$|^[a-f0-9]{64}$)."
  }
```
It the digest is corrupted you do not get any help. I think that data that users can paste in GUI should be handled by RPCAPI in such a way that a meaningful error message can be returned. Of course, the GUI must also be prepared to display it for the user.

Another possibility of error in digest is that some digit is replaced. In that case the test is executed and there is some meaningful error messages for the user.

Reply:
@mattias-p, any comments?

Also see #789

Reply:
> @pnax @matsduf Could you guys go back through your review comments above and "Resolve conversions" that you think are resolved? I believe I've addressed all of them except the MySQL IPv6 issue.

I have done so, and there are two places where I would like to see a reply from you.

Reply:
I agree that regexes don't make for great validation errors. But I don't think this PR is the place to fix it. Maybe you could create an issue for it?

Reply:
See #789